core/src/main/java/io/questdb/cairo/mv/MatViewRefreshJob.java

```diff
@@ -209,23 +222,36 @@ private SampleByIntervalIterator findSampleByIntervals(
             minTs = txnRangeLoader.getMinTimestamp();
             maxTs = txnRangeLoader.getMaxTimestamp();
             // Check if refresh limit should be applied.
-            try (TableMetadata matViewMeta = engine.getTableMetadata(viewToken)) {
-                final int limit = matViewMeta.getMatViewRefreshLimitHoursOrMonths();
-                if (limit != 0) {
-                    final long now = microsecondClock.getTicks();
-                    if (limit > 0) { // hours
-                        minTs = Math.max(minTs, now - Timestamps.HOUR_MICROS * limit);
-                    } else { // months
-                        minTs = Math.max(minTs, Timestamps.addMonths(now, -limit));
-                    }
+            if (viewTableMeta.refreshLimit != 0) {
+                if (viewTableMeta.refreshLimit > 0) { // hours
+                    minTs = Math.max(minTs, now - Timestamps.HOUR_MICROS * viewTableMeta.refreshLimit);
+                } else { // months
+                    minTs = Math.max(minTs, Timestamps.addMonths(now, -viewTableMeta.refreshLimit));
                 }
             }
-        } else if (rangeFrom != Numbers.LONG_NULL && rangeTo != Numbers.LONG_NULL) {
+        } else if (rangeRefresh) {
             // Range refresh.
-            // Consider actual min/max timestamps in the table data to avoid redundant
-            // query executions.
-            minTs = Math.max(rangeFrom, baseTableReader.getMinTimestamp());
-            maxTs = Math.min(rangeTo, baseTableReader.getMaxTimestamp());
+            if (rangeFrom == Numbers.LONG_NULL) {
+                // Timer-triggered period refresh.
+                long periodLo = viewState.getLastPeriodHi();
```

Comment:
For some reason, last period hi may not get updated (or read from disk) after a server restart.